cpp/oneapi/dal/algo/pca/backend/gpu/infer_kernel_dpc.cpp

```diff
@@ -26,39 +26,162 @@
+
+    auto centered_data_ptr = centered_data.get_mutable_data();
+    auto means_ptr = means.get_data();
+
+    auto centered_event = q.submit([&](sycl::handler& h) {
+        const auto range = bk::make_range_2d(row_count, column_count);
+        h.depends_on(copy_event);
+        h.parallel_for(range, [=](sycl::id<2> id) {
+            const std::int64_t i = id[0];
+            const std::int64_t j = id[1];
+            centered_data_ptr[i * column_count + j] =
+                centered_data_ptr[i * column_count + j] - means_ptr[j];
+        });
+    });
+    return std::make_tuple(centered_data, centered_event);
+}
+
+template <typename Float>
+auto get_scaled(sycl::queue& q,
+                const pr::ndview<Float, 2>& data,
+                const pr::ndview<Float, 1>& variances,
+                const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_scaled_data, q);
+    const std::int64_t row_count = data.get_dimension(0);
+    const std::int64_t column_count = data.get_dimension(1);
+
+    auto scaled_data = pr::ndarray<Float, 2>::empty(q, { row_count, column_count }, alloc::device);
+    auto copy_event = copy(q, scaled_data, data, { deps });
+
+    auto scaled_data_ptr = scaled_data.get_mutable_data();
+    auto variances_ptr = variances.get_data();
+
+    auto scaled_event = q.submit([&](sycl::handler& h) {
+        const auto range = bk::make_range_2d(row_count, column_count);
+        h.depends_on(copy_event);
+        h.parallel_for(range, [=](sycl::id<2> id) {
+            const std::int64_t i = id[0];
+            const std::int64_t j = id[1];
+            scaled_data_ptr[i * column_count + j] =
+                scaled_data_ptr[i * column_count + j] * (1 / sqrt(variances_ptr[j]));
+        });
+    });
+    return std::make_tuple(scaled_data, scaled_event);
+}
+
+template <typename Float>
+auto get_whitened(sycl::queue& q,
+                  const pr::ndview<Float, 2>& data,
+                  const pr::ndview<Float, 1>& eigenvalues,
+                  const bk::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(compute_whitened_data, q);
+    const std::int64_t row_count = data.get_dimension(0);
+    const std::int64_t column_count = data.get_dimension(1);
+
+    auto whitened_data =
+        pr::ndarray<Float, 2>::empty(q, { row_count, column_count }, alloc::device);
+    auto copy_event = copy(q, whitened_data, data, { deps });
+
+    auto whitened_data_ptr = whitened_data.get_mutable_data();
+    auto eigenvalues_ptr = eigenvalues.get_data();
+
+    auto whitened_event = q.submit([&](sycl::handler& h) {
+        const auto range = bk::make_range_2d(row_count, column_count);
+        h.depends_on(copy_event);
+        h.parallel_for(range, [=](sycl::id<2> id) {
+            const std::int64_t i = id[0];
+            const std::int64_t j = id[1];
+            whitened_data_ptr[i * column_count + j] =
+                whitened_data_ptr[i * column_count + j] * (1 / sqrt(eigenvalues_ptr[j]));
```

Comment:
```suggestion
                whitened_data_ptr[i * column_count + j] * invsqrt(eigenvalues_ptr[j]);
```

Reply:
Another thought about possible performance improvements:
In this version of the code `invsqrt` is computed `row_count * column_count` times. Maybe it would be beneficial to pre-compute a vector of invsqrt(eigenvalues_ptr), which involves just `column_count` calculations of invsqrt. [oneMKL's VM library can be used to perform this]. And then use the pre-computed data in whitening operation.